### PR TITLE
Fix for Android - Dynamic Updates to CollectionView Header/Footer and Templates Are Not Displayed 

### LINKED ISSUE #28676: Android - CollectionView Fails to Update Group Header/Footer Template Assigned at Runtime
### Description

- An unintended synchronization exists between the GroupHeaderTemplate / GroupFooterTemplate and the Header / Footer views in .NET MAUI CollectionView. When setting a header or footer template, the corresponding header or footer view is also updated unexpectedly, and vice versa.
- The CollectionView fails to reflect dynamically updated Header, Footer, or their templates when modified at runtime. As a result, newly assigned headers, footers, or templates do not render correctly, leading to missing UI elements.

### Actual Issue in Android

https://github.com/user-attachments/assets/aba530ae-8aff-446d-bee3-d1914c5adc61

### Expected

https://github.com/user-attachments/assets/d369b746-8fc8-45fc-a70e-e3f60320a8f4

### Steps to Reproduce

- Launch the Application on Android - [GroupingSample.zip](https://github.com/user-attachments/files/19499219/GroupingSample.zip)
- Notice the issue based on the shared video reference.

### Link to public reproduction project repository

_No response_

### Version with bug

9.0.50 SR5

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android

### Did you find any workaround?

_No response_

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by SuthiYuvaraj**
### Issue Description:
 
- Dynamic updates to CollectionView's Header/Footer (both text and views) are not adequately reflected on Android.
- Changes to `HeaderTemplate` and `FooterTemplate` are also not applied correctly at runtime.
 
### Root Cause:
 
- `ItemsViewPropertyChanged` did not handle updates for `HeaderTemplateProperty` and `FooterTemplateProperty`.
- `OnBindViewHolder` did not account for rendering updated header/footer content or templates.
- Templated headers and footers were not properly managed when the content changed dynamically.
 
### Description of Change
 
- Added support for `HeaderTemplateProperty` and `FooterTemplateProperty` in the `ItemsViewPropertyChanged` method to ensure runtime changes are detected and processed.
- Updated the `OnBindViewHolder` method to handle both simple views and string-based headers/footers using SimpleViewHolder.
- Bind the correct templates when a header or footer is templated.
- Enhanced `BindTemplatedItemViewHolder` to correctly create and bind TemplatedItemViewHolder for HeaderTemplate and FooterTemplate, ensuring updates are accurately reflected in the UI.
 
### Issues Fixed
Fixes #28676
 
### Tested the behaviour on the following platforms
- [x] Android
- [x] Windows
- [x] iOS
- [x] Mac
 
### Output Screenshot
Before Issue Fix | After Issue Fix |
|----------|----------|
|<video width="250" height="150" alt="Before Fix" src="https://github.com/user-attachments/assets/72627eed-d6c2-4625-99ed-38e4dbd90742">|<video width="250" height="150" alt="After Fix" src="https://github.com/user-attachments/assets/0a9fc95d-2c10-439c-bdb6-b315f05e6113">|

### CODE REVIEW

File: src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue28676.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	public void CollectionViewHeaderViewShouldChangeDynamically()
+	{
+		App.WaitForElement("UpdatedCollectionViewHeader");
+		App.WaitForElement("Issue28676ChangeHeaderView");
+		App.Tap("Issue28676ChangeHeaderView");
+		App.WaitForElement("Updated HeaderView");
+	}
+
+	[Test, Order(4)]
+	public void CollectionViewFooterViewShouldChangeDynamically()
```
Comment (jsuarezruiz) [inline src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue28676.cs:42]: This tests are failing on iOS:

image

 ```
  at UITest.Appium.HelperExtensions.Wait(Func`1 query, Func`2 satisfactory, String timeoutMessage, Nullable`1 timeout, Nullable`1 retryFrequency) in /_/src/TestUtils/src/UITest.Appium/HelperExtensions.cs:line 2404
   at UITest.Appium.HelperExtensions.WaitForAtLeastOne(Func`1 query, String timeoutMessage, Nullable`1 timeout, Nullable`1 retryFrequency) in /_/src/TestUtils/src/UITest.Appium/HelperExtensions.cs:line 2421
   at UITest.Appium.HelperExtensions.WaitForElement(IApp app, String marked, String timeoutMessage, Nullable`1 timeout, Nullable`1 retryFrequency, Nullable`1 postTimeout) in /_/src/TestUtils/src/UITest.Appium/HelperExtensions.cs:line 665
   at Microsoft.Maui.TestCases.Tests.Issues.Issue28676.CollectionViewFooterViewShouldChangeDynamically() in /_/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue28676.cs:line 47
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

File: src/Controls/src/Core/Handlers/Items/Android/Adapters/StructuredItemsViewAdapter.cs [hunk 2]
```diff
@@ -33,12 +33,13 @@ protected override void ItemsViewPropertyChanged(object sender, PropertyChangedE
 		{
 			base.ItemsViewPropertyChanged(sender, property);
 
-			if (property.Is(Microsoft.Maui.Controls.StructuredItemsView.HeaderProperty))
+			if (property.Is(Microsoft.Maui.Controls.StructuredItemsView.HeaderProperty) || property.Is(Microsoft.Maui.Controls.StructuredItemsView.HeaderTemplateProperty))
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/Handlers/Items/Android/Adapters/StructuredItemsViewAdapter.cs:36]: The test `HeaderFooterTemplateWorks` is failing on Android, could you verify if is related with the changes?

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]

**Review by sheiksyedm — CHANGES_REQUESTED** [review]
https://github.com/dotnet/maui/pull/28904#discussion_r2053385925

**Comment by SuthiYuvaraj** [comment]
Hi @jsuarezruiz  , The failures in the HeaderFooterTemplateWorks test on Android and the above mentioned testcases on iOS do not appear to be related to my changes. This is unrelated to my fix , Kindly rerun the CI again

**Review by kubaflo — CHANGES_REQUESTED** [review]
Could you please review the AI's summary?

File: src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs [hunk 3]
```diff
@@ -14,16 +14,86 @@ public partial class StructuredItemsViewHandler<TItemsView> : ItemsViewHandler<T
...
+			}
+			else if (hasHeaderOrFooter != exists)
+			{
+				recyclerView.UpdateAdapter();
+			}
+		}
+
+		bool DoesHeaderExist(RecyclerView.Adapter adapter)
+		{
+			if (adapter.ItemCount >= 0)
```
Comment (kubaflo) [inline src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs]: Could you please check?

**Review by kubaflo — CHANGES_REQUESTED** [review]

**Review by kubaflo — CHANGES_REQUESTED** [review]
Could you please review the ai's summary?

**Review by MauiBot — DISMISSED** [review]
🤖 **Automated review — alternative fix proposed**

The expert-reviewer evaluation compared the PR fix against #4 automatically generated candidates and selected `try-fix-4` as the strongest fix.

**Why:** try-fix-4 provides a 2-line unconditional UpdateAdapter() fix in MapHeaderTemplate/MapFooterTemplate that exactly mirrors the existing MapItemSizingStrategy pattern. It correctly handles all scenarios including empty item sources (where the PR's DoesHeaderExist/DoesFooterExist methods have a logic error with EmptyViewAdapter), requires no complex view-type probing or bare catch blocks, and passes all 3 tests. Of the 5 passing alternatives and the pr-plus-reviewer candidate, try-fix-4 is the simplest, most maintainable, and bug-free solution.

Please consider applying the candidate diff below (or use it as guidance). Once you push an update, this workflow will re-trigger and re-evaluate.

<details><summary>Candidate diff (`try-fix-4`)</summary>

```diff
diff --git a/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs b/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs
index 109892b41c..4c73598be1 100644
--- a/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs
+++ b/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs
@@ -14,10 +14,12 @@ namespace Microsoft.Maui.Controls.Handlers.Items
 
 		public static void MapHeaderTemplate(StructuredItemsViewHandler<TItemsView> handler, StructuredItemsView itemsView)
 		{
+			(handler.PlatformView as IMauiRecyclerView<TItemsView>)?.UpdateAdapter();
 		}
 
 		public static void MapFooterTemplate(StructuredItemsViewHandler<TItemsView> handler, StructuredItemsView itemsView)
 		{
+			(handler.PlatformView as IMauiRecyclerView<TItemsView>)?.UpdateAdapter();
 		}
 
 		public static void MapItemsLayout(StructuredItemsViewHandler<TItemsView> handler, StructuredItemsView itemsView)
diff --git a/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue28509.cs b/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue28509.cs
index 5c7ff7d28f..66f59968a6 100644
--- a/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue28509.cs
+++ b/src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue28509.cs
@@ -1,5 +1,3 @@
-#if TEST_FAILS_ON_ANDROID
-// https://github.com/dotnet/maui/issues/28676
 using NUnit.Framework;
 using UITest.Appium;
 using UITest.Core;
@@ -53,6 +51,6 @@ public class Issue28509 : _IssuesUITest
 		App.WaitForElement("Header Changed");
 		App.Click("ToggleFooterButton");
 		App.WaitForElement("Footer Changed");
+		App.TapBackArrow();
 	}
-}
-#endif
\ No newline at end of file
+}
\ No newline at end of file
```

</details>

**Review by kubaflo — CHANGES_REQUESTED** [review]
Could you please review AI's suggestions?

**Comment by SuthiYuvaraj** [comment]
Hi @kubaflo , I reviewed the AI‑suggested approach, but it requires a full adapter refresh. This would cause the RecyclerView to rebind all visible items and re-evaluate view types, including headers and footers. That introduces unnecessary overhead and redundant calls, so I prefer to proceed with the existing code changes rather than invoking updateAdapter().

**Review by MauiBot — CHANGES_REQUESTED** [review]
🤖 **Automated review — alternative fix proposed**

The expert-reviewer evaluation compared the PR fix against #2 automatically generated candidates and selected `try-fix-2` as the strongest fix.

**Why:** try-fix-2 (claude-sonnet-4.6) wins: it is the simplest correct fix — just 2 lines added to MapHeaderTemplate/MapFooterTemplate in the handler (mirroring existing MapItemSizingStrategy pattern), plus 9 lines removed from the now-redundant adapter handler. It passes all 3 gate tests, has no EmptyViewAdapter edge-case bug, requires no PublicAPI changes, and is net -11 lines vs the PR's +70. The PR fix works but introduces unnecessarily complex detection logic with a real correctness bug when EmptyViewAdapter is active.

Please consider applying the candidate diff below (or use it as guidance). Once you push an update, this workflow will re-trigger and re-evaluate.

<details><summary>Candidate diff (`try-fix-2`)</summary>

```diff
diff --git a/src/Controls/src/Core/Handlers/Items/Android/Adapters/StructuredItemsViewAdapter.cs b/src/Controls/src/Core/Handlers/Items/Android/Adapters/StructuredItemsViewAdapter.cs
index 42c8055d9a..2fafc35a97 100644
--- a/src/Controls/src/Core/Handlers/Items/Android/Adapters/StructuredItemsViewAdapter.cs
+++ b/src/Controls/src/Core/Handlers/Items/Android/Adapters/StructuredItemsViewAdapter.cs
@@ -33,16 +33,8 @@ namespace Microsoft.Maui.Controls.Handlers.Items
 		{
 			base.ItemsViewPropertyChanged(sender, property);
 
-			if (property.Is(Microsoft.Maui.Controls.StructuredItemsView.HeaderProperty) || property.Is(Microsoft.Maui.Controls.StructuredItemsView.HeaderTemplateProperty))
-			{
-				UpdateHasHeader();
-				NotifyDataSetChanged();
-			}
-			else if (property.Is(Microsoft.Maui.Controls.StructuredItemsView.FooterProperty) || property.Is(Microsoft.Maui.Controls.StructuredItemsView.FooterTemplateProperty))
-			{
-				UpdateHasFooter();
-				NotifyDataSetChanged();
-			}
+			// Header/Footer updates are handled by the handler mapper (MapHeaderTemplate/MapFooterTemplate)
+			// via UpdateAdapter(), which creates a fresh adapter with the correct header/footer state.
 		}
 
 		public override int GetItemViewType(int position)
diff --git a/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs b/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs
index 109892b41c..af269d7d11 100644
--- a/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs
+++ b/src/Controls/src/Core/Handlers/Items/StructuredItemsViewHandler.Android.cs
@@ -13,12 +13,10 @@ namespace Microsoft.Maui.Controls.Handlers.Items
 		protected override StructuredItemsViewAdapter<TItemsView, IItemsViewSource> CreateAdapter() => new(VirtualView);
 
 		public static void MapHeaderTemplate(StructuredItemsViewHandler<TItemsView> handler, StructuredItemsView itemsView)
-		{
-		}
+			=> (handler.PlatformView as IMauiRecyclerView<TItemsView>)?.UpdateAdapter();
 
 		public static void MapFooterTemplate(StructuredItemsViewHandler<TItemsView> handler, StructuredItemsView itemsView)
-		{
-		}
+			=> (handler.PlatformView as IMauiRecyclerView<TItemsView>)?.UpdateAdapter();
 
 		public static void MapItemsLayout(StructuredItemsViewHandler<TItemsView> handler, StructuredItemsView itemsView)
 		{

```

</details>

**Review by kubaflo — APPROVED** [review]
